File: Eunleelee/1월 3일/뒤에_있는_큰_수_찾기.py
def solution(numbers):
    # numbers의 길이만큼 값 -1을 담은 배열
    answer = [-1] * len(numbers)
    stack = []

    for index, value in enumerate(numbers):
        # 배열 stack에 값이 존재하고 numbers[stack[-1]] 값이 현재 비교하는 값보다 작으면
        while stack and numbers[stack[-1]] < value:
            # stack.pop이 되는 값을 인덱스로 answer에 현재 값을 저장
            answer[stack.pop()] = value

        stack.append(index)

    return answer

print(solution([9, 1, 5, 3, 6, 2]))

File: yht0827/1월/1월 3일/뒤에 있는 큰수 찾기.java
import java.util.Stack;

class Solution {
	public static int[] solution(int[] numbers) {
		int[] answer = new int[numbers.length];
		Stack<Integer> stack = new Stack<>();
		for (int i = numbers.length - 1; i >= 0; i--) {
			while (!stack.isEmpty()) {
				int num = stack.peek();
				if (num > numbers[i]) {
					answer[i] = num;
					break;
				} else {
					stack.pop();
				}
			}
			if (stack.isEmpty()) {
				answer[i] = -1;
			}

			stack.push(numbers[i]);
		}
		return answer;
	}
}